File: agenkit-java/src/main/java/io/agenkit/memory/EphemeralMemory.java
package io.agenkit.memory;

import io.agenkit.core.Message;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * In-memory storage that does not persist across sessions.
 */
public final class EphemeralMemory implements Memory {

    private final List<Message> messages = new ArrayList<>();
    private final int maxSize;

    public EphemeralMemory(int maxSize) {
        this.maxSize = maxSize;
    }

    public EphemeralMemory() {
        this(1000);
    }

    @Override
    public synchronized void store(Message message) {
        messages.add(message);
        if (messages.size() > maxSize) {
            messages.remove(0);
        }
    }

    @Override
    public synchronized List<Message> retrieve(String query, int topK) {
        // Simple recency-based retrieval
        int start = Math.max(0, messages.size() - topK);
        return Collections.unmodifiableList(new ArrayList<>(messages.subList(start, messages.size())));
    }

    @Override
    public synchronized int size() { return messages.size(); }

    @Override
    public synchronized void clear() { messages.clear(); }
}

File: agenkit-scala/src/main/scala/io/agenkit/memory/EphemeralMemory.scala
package io.agenkit.memory

import io.agenkit.core.Message

/** In-memory storage with a bounded capacity; oldest entries are evicted first. */
class EphemeralMemory(
  maxSize: Int = 1000
) extends Memory:
  private val items = collection.mutable.ListBuffer[Message]()

  def store(message: Message): Unit =
    items.synchronized {
      items += message
      if items.size > maxSize then items.remove(0)
    }

  def retrieve(query: String, limit: Int = 10): List[Message] =
    items.synchronized {
      val q = query.toLowerCase
      items.filter(m => m.contentString.toLowerCase.contains(q))
        .takeRight(limit)
        .toList
    }

  def clear(): Unit = items.synchronized { items.clear() }
  def size: Int     = items.synchronized { items.size }
